File: AdventOfCode/2023/day2.cpp
#include <bits/stdc++.h>
using namespace std;

namespace util 
{
vector<string> split(string s, string delimiter) 
{
    size_t pos_start = 0, pos_end, delim_len = delimiter.length();
    string token;
    vector<string> res;

    while ((pos_end = s.find(delimiter, pos_start)) != string::npos) {
        token = s.substr(pos_start, pos_end - pos_start);
        pos_start = pos_end + delim_len;
        res.push_back(token);
    }

    res.push_back (s.substr(pos_start));
    return res;
}
}

namespace 
{

void computeMaximums(const string& set, int& maxRed, int& maxBlue, int& maxGreen)
{
    const vector<string> cubeValuePairs = util::split(set, ", ");
    for (const auto& cubeValuePair : cubeValuePairs) {
        const auto cubeValueSplit = util::split(cubeValuePair, " ");
        int value = stoi(cubeValueSplit[0]);  
        const string cube = cubeValueSplit[1];

        if (cube == "red") maxRed = max(maxRed, value);
        else if (cube == "blue") maxBlue = max(maxBlue, value);
        else if (cube == "green") maxGreen = max(maxGreen, value);
    }
}

bool is_set_possible(const string& set)
{
    int redValue = 0, blueValue = 0, greenValue = 0;

    const vector<string> cubeValuePairs = util::split(set, ", ");
    for (const auto& cubeValuePair : cubeValuePairs) {
        const auto cubeValueSplit = util::split(cubeValuePair, " ");
        int value = stoi(cubeValueSplit[0]);  
        const string cube = cubeValueSplit[1];

        if (cube == "red") redValue += value;
        else if (cube == "blue") blueValue += value;
        else if (cube == "green") greenValue += value;
    }

    return !(redValue > 12 || blueValue > 14 || greenValue > 13);
}
}

int computePartA(string& s) 
{  
    vector<string> split = util::split(s, ": ");
    const auto gameIdSplit = util::split(split[0], " ");
    int gameId = stoi(gameIdSplit[1]);
    bool isGamePossible = true;

    const auto sets = util::split(split[1], "; ");
    for (const auto& set : sets) {
        if (!is_set_possible(set)) {
            isGamePossible = false;
            break;
        }
    }

    return isGamePossible ? gameId : 0;
}

int computePartB(string& s) 
{  
    vector<string> split = util::split(s, ": ");
    const auto gameIdSplit = util::split(split[0], " ");
    const auto sets = util::split(split[1], "; ");

    int maxRed = 0, maxBlue = 0, maxGreen = 0;

    for (const string& set : sets) {
        computeMaximums(set, maxRed, maxBlue, maxGreen);
    }

    return maxBlue * maxRed * maxGreen;
}


int main()
{
    int totA = 0, totB = 0;
    while (true)
    {
        string s;
        getline(cin, s);
        if (s.empty()) break;

        totA += computePartA(s);
        totB += computePartB(s);
    }
    cout << totA << ", " << totB << endl;
    return 0;
}

File: AdventOfCode/2023/day2.rs
use std::fs::read_to_string;

fn is_set_possible(set: Vec<&str>) -> bool {
    let cube_value_pairs: Vec<&str> = set.split(", ").collect();
    let mut red_value = 0;
    let mut blue_value = 0;
    let mut green_value = 0;

    for cube_value_pair in cube_value_pairs.iter() {
        let cube_value_split: Vec<&str> = cube_value_pair.split_whitespace().collect();
        let value = cube_value_split[0].parse::<i32>().unwrap();
        match cube_value_split[1] {
            "red" => red_value += value,
            "blue" => blue_value += value,
            "green" => green_value += value,
            _ => {}
        }
    }

    !(red_value > 12 || blue_value > 14 || green_value > 13)
}

fn solution_day2a(filename: &str) -> i32 {
    let mut result = 0;

    for line in read_to_string(filename).unwrap().lines() {
        let split: Vec<&str> = line.split(": ").collect();
        let game_id = split[0].split_whitespace().collect::<Vec<_>>()[1]
            .parse::<i32>()
            .unwrap();
        let mut is_game_possible = true;

        let sets: Vec<&str> = split[1].split("; ").collect();
        for set in sets.iter() {
            if (!is_set_possible(set)) {
                is_game_possible = false;
                break;
            }
        }

        if is_game_possible {
            result += game_id;
        }
    }
    result
}

fn solution_day2b(filename: &str) -> i32 {
    let mut result = 0;

    for line in read_to_string(filename).unwrap().lines() {
        let mut max_red = 0;
        let mut max_blue = 0;
        let mut max_green = 0;

        let tries: Vec<&str> = line.split(": ").collect::<Vec<_>>()[1]
            .split("; ")
            .collect();
        for local_try in tries.iter() {
            let cubes: Vec<&str> = local_try.split(", ").collect();

            for cube in cubes.iter() {
                let inner_split: Vec<&str> = cube.split_whitespace().collect();
                let val = inner_split[0].parse::<i32>().unwrap();
                match inner_split[1] {
                    "red" => max_red = std::cmp::max(max_red, val),
                    "blue" => max_blue = std::cmp::max(max_blue, val),
                    "green" => max_green = std::cmp::max(max_green, val),
                    _ => {}
                }
            }
        }

        result += (max_red * max_blue * max_green);
    }
    result
}

fn main_entry() {
    println!("{}", solution_day2a("src/day2_input.txt"));
    println!("{}", solution_day2b("src/day2_input.txt"));
}
